### PR TITLE
8301562: [Lilliput] Runtime flag to enable Lilliput

### PR DESCRIPTION
**Pull request by rkennke**
This change introduces an experimental runtime flag -XX:[+/-]UseCompactObjectHeaders which can be used to disable Lilliput. The flag guards all the relevant places that directly affect the object layout and the accesses to Klass* etc. The change also restores the previous behaviour when -UseCompactObjectHeaders is selected.

One special difficulty is that CDS archives that are generated with compact object headers cannot currently be read with old-style headers, and vice versa. I work around this by disabling CDS when UseCompactObjectHeaders is set to non-default. This makes a few tests fail, but should otherwise be harmless. (In theory, we could implement parsing archives and loading objects with the correct header layout, but ... life's too short ;-) )

Testing:
 - [x] tier1 (x86_64, x86_32, aarch64) +UseCompactObjectHeaders
 - [x] tier2 (x86_64, x86_32, aarch64) +UseCompactObjectHeaders
 - [x] tier1 (x86_64, x86_32, aarch64) -UseCompactObjectHeaders
 - [x] tier2 (x86_64, x86_32, aarch64) -UseCompactObjectHeaders

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8301562](https://bugs.openjdk.org/browse/JDK-8301562): [Lilliput] Runtime flag to enable Lilliput

### Reviewers
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - Committer) ⚠️ Review applies to [aebedb63](https://git.openjdk.org/lilliput/pull/70/files/aebedb63e0968d5f8ed2a0a777f69b3bfbd464d7)

### Contributors
 * Thomas Stuefe `<stuefe@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/lilliput pull/70/head:pull/70` \
`$ git checkout pull/70`

Update a local copy of the PR: \
`$ git checkout pull/70` \
`$ git pull https://git.openjdk.org/lilliput pull/70/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 70`

View PR using the GUI difftool: \
`$ git pr show -t 70`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/lilliput/pull/70.diff">https://git.openjdk.org/lilliput/pull/70.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/globals.hpp [hunk 1]
```diff
@@ -125,10 +125,13 @@ const size_t minimumSymbolTableSize = 1024;
           "Use 32-bit object references in 64-bit VM. "                     \
           "lp64_product means flag is always constant in 32 bit VM")        \
                                                                             \
-  product(bool, UseCompressedClassPointers, true,                           \
+  product(bool, UseCompressedClassPointers, false,                          \
```
Comment (tstuefe) [inline src/hotspot/share/runtime/globals.hpp:128]: this is just temporary, for testing, right?

Comment (tstuefe) [inline src/hotspot/share/runtime/globals.hpp:128]: just temporary, for testing, right?

**Comment by tstuefe** [comment]
LogKlassAlignmentInBytes and everything depending on it will have to be switched to runtime-controlled variables. Nothing complex, just annoying. I'm on it.

**Comment by tstuefe** [comment]
Hi Roman,

Some fixes from my side, see branch https://github.com/tstuefe/lilliput/commits/fixes-to-usecompactobjectheader

Suggested changes:

1) https://github.com/tstuefe/lilliput/commit/38dcba1d06a56b01cf7c9a593656f1ae5109b9b0

this makes KlassAlignmentInBytes and friends runtime-controlled. This is really ugly, but the whole ccs- and compressed klass pointer encoding initialization deserves an overhaul, and this is the least intrusive fix possible.

2) https://github.com/tstuefe/lilliput/commit/ad929fafbb4f7c5785e5627b53f0de33dafd2501

Needed to fix a crash when compact headers are enabled.

3) https://github.com/tstuefe/lilliput/commit/e854428c077e596f81e1349107bc37c2b80c5237

Expands gtests to also test all permutations of CCS and CompactHeaders

4) https://github.com/tstuefe/lilliput/commit/4c01e6a017bda54e6cdee62ef8814756829fb489

Needed to move constraint check for CompressedClassPointerSize to after ergo

Cheers, thomas

**Comment by rkennke** [comment]
@tstuefe thanks for you contributions! Unfortunately, I am still getting failures when building on aarch64:

Optimizing the exploded image
# To suppress the following error report, specify this argument
# after -XX: or in .hotspotrc:  SuppressErrorAt=/metaspace.cpp:621
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  Internal Error (/home/rkennke/src/openjdk/lilliput/src/hotspot/share/memory/metaspace.cpp:621), pid=10318, tid=10323
#  assert(CompressedKlassPointers::is_valid_base(a)) failed: Sanity
#
# JRE version:  (21.0) (fastdebug build )
# Java VM: OpenJDK 64-Bit Server VM (fastdebug 21-internal-adhoc.rkennke.lilliput, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# V  [libjvm.so+0x12f2050]  Metaspace::reserve_address_space_for_compressed_classes(unsigned long)+0x190
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /local/home/rkennke/src/openjdk/lilliput/make/hs_err_pid10318.log
#
#
/usr/bin/bash: line 1: 10318 Aborted                 /home/rkennke/src/openjdk/lilliput/build/linux-aarch64-server-fastdebug/jdk/bin/java -Xms64M -Xmx1600M -cp /home/rkennke/src/openjdk/lilliput/build/linux-aarch64-server-fastdebug/buildtools/tools_jigsaw_classes --add-exports java.base/jdk.internal.module=ALL-UNNAMED build.tools.jigsaw.AddPackagesAttribute /home/rkennke/src/openjdk/lilliput/build/linux-aarch64-server-fastdebug/jdk > >(/usr/bin/tee -a /home/rkennke/src/openjdk/lilliput/build/linux-aarch64-server-fastdebug/jdk/_optimize_image_exec.log) 2> >(/usr/bin/tee -a /home/rkennke/src/openjdk/lilliput/build/linux-aarch64-server-fastdebug/jdk/_optimize_image_exec.log >&2)
gmake[3]: *** [/home/rkennke/src/openjdk/lilliput/build/linux-aarch64-server-fastdebug/jdk/_optimize_image_exec.marker] Error 1
gmake[2]: *** [exploded-image-optimize] Error 1

ERROR: Build failed for target 'images' in configuration 'linux-aarch64-server-fastdebug' (exit code 2)

**Comment by tstuefe** [comment]
Yeah, 32-bit is also broken. Sigh.

**Comment by tstuefe** [comment]
Okay, some more fixes:

https://github.com/tstuefe/lilliput/commit/b2f0364062e706303dc6ec8ee434700a51a99e30 : fix up aarch64 CCS allocation. 

I have planned to overhaul CCS initialization for a while, maybe its time. There are conceptional flaws, and it is really ugly too with all that aarch64-specific knowledge hardcoded into shared metaspace coding, and ppc64 piggybacking atop of that :-(

https://github.com/tstuefe/lilliput/commit/37d386aa1653040859442c0926753a78c17cdd84 : obvious fix for c1 on aarch64

https://github.com/tstuefe/lilliput/commit/dd5af41ce8845fc56f246321cdff55343671bcad : bunch of build fixes on x86. This is another sore point, I really dislike all that 64-bit specific coding compiled into x86. It is this way in head too. I would love to see this fixed.

Remaining, as far as I see:

- `-XX:-UseCompressedClassPointers` crashes on aarch64 with `Internal Error (/shared/projects/openjdk/lilliput/source/src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:2570)`
- there are more fixes to be done for x86. `LIR_Assembler::emit_arraycopy` does not build.

**Comment by tstuefe** [comment]
P.s. I'm out of time now until next month (probably).

**Comment by shipilev** [comment]
I can take a look at x86_32 stuff, if you need it.

**Comment by tstuefe** [comment]
Taking a look. Tell me if you need help merging JDK21+12, I removed MetaspaceReclaimPolicy=none upstream

**Pull request closed without merging**